lib/BuildSystem/BuildSystem.cpp

```diff
@@ -524,7 +524,18 @@ class CommandTask : public Task {
   }
 
   virtual void inputsAvailable(BuildEngine& engine) override {
-    command.inputsAvailable(getBuildSystem(engine).getCommandInterface(), this);
+    bool shouldSkip = !getBuildSystem(engine).getCommandInterface()
+      .getDelegate().shouldCommandStart(&command);
+
+    if (shouldSkip) {
+      getBuildSystem(engine).getCommandInterface().taskIsComplete(
+          this, BuildValue::makeSkippedCommand());
+      getBuildSystem(engine).getCommandInterface().getDelegate()
+        .commandFinished(&command);
```

Comment:
The delegate doc comments guarantee that commandFinished is called after commandPreparing, and commandPreparing has been called by the time shouldCommandStart is called. It's a little confusing because the term Finished is naturally paired with Started, yet Preparing and ShouldStart guarantee Finished. I wonder if Preparing and ShouldStart should guarantee Finished at all? Or if the delegate does need an end callback to match Preparing, maybe it should be something new? Thoughts?